# Apply linear Regression

In [45]:
from sklearn.datasets import load_diabetes

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import numpy as np
import pandas as pd

## load data 

In [46]:
df = load_diabetes(as_frame=True)
df['data']

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
0,0.038076,0.050680,0.061696,0.021872,-0.044223,-0.034821,-0.043401,-0.002592,0.019907,-0.017646
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204
2,0.085299,0.050680,0.044451,-0.005670,-0.045599,-0.034194,-0.032356,-0.002592,0.002861,-0.025930
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362
4,0.005383,-0.044642,-0.036385,0.021872,0.003935,0.015596,0.008142,-0.002592,-0.031988,-0.046641
...,...,...,...,...,...,...,...,...,...,...
437,0.041708,0.050680,0.019662,0.059744,-0.005697,-0.002566,-0.028674,-0.002592,0.031193,0.007207
438,-0.005515,0.050680,-0.015906,-0.067642,0.049341,0.079165,-0.028674,0.034309,-0.018114,0.044485
439,0.041708,0.050680,-0.015906,0.017293,-0.037344,-0.013840,-0.024993,-0.011080,-0.046883,0.015491
440,-0.045472,-0.044642,0.039062,0.001215,0.016318,0.015283,-0.028674,0.026560,0.044529,-0.025930


In [47]:
df['target']

0      151.0
1       75.0
2      141.0
3      206.0
4      135.0
       ...  
437    178.0
438    104.0
439    132.0
440    220.0
441     57.0
Name: target, Length: 442, dtype: float64

In [48]:
# import as numpy arrays
X,y = load_diabetes(return_X_y=True)

X = input data     
y = output data

In [49]:
print(X.shape)
print(y.shape)

(442, 10)
(442,)


In [50]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=2)

In [51]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(353, 10)
(353,)
(89, 10)
(89,)


## apply OLS

In [52]:
# fit the model on training data
reg_model = LinearRegression()
reg_model.fit(X_train,y_train)

LinearRegression()

## find parameters

In [53]:
# after fitting models print the parameters 
# we apply the multiple linear regression so 1 intercept and 10 coefficent
print('intercept:',reg_model.intercept_)
print('coefficents:', reg_model.coef_)

intercept: 151.88331005254167
coefficents: [  -9.15865318 -205.45432163  516.69374454  340.61999905 -895.5520019
  561.22067904  153.89310954  126.73139688  861.12700152   52.42112238]


so learning from the trainining data = samples          
so predication in the testing data = population = unseen data 

In [54]:
# predication of the model 
y_pred = reg_model.predict(X_test)

## r2 score

In [55]:
# now compare how good is our predication 
# error = actual - predication 
r2_score(y_test,y_pred)

0.439933866156897

now 43 % varience is expalined by data lakin phichali baar 45 tha . reason is sampling . phicha baar alaga sample ka data tha is par alag sample data ha 

# Our Stochastic gradient class

so use the same code from batch gradient descant 

In [56]:
class GradientDescantRegressor:

    def __init__(self,learning_rate=0.001,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept_ = 0
        self.coef_ = np.ones(X_train.shape[1])
        self.coef_ = np.ones(X_train.shape[1])
        
        for i in range(self.epoches):
            # update all the coeff and the intercept
            y_hat = np.dot(X_train,self.coef_) + self.intercept_
            slope_of_intercept  = -2 * np.mean(y_train - y_hat)
            self.intercept_ = self.intercept_- (self.learning_rate * slope_of_intercept) 

            slope_of_coefficents  =  -2 * np.dot((y_train - y_hat),X_train)/X_train.shape[0]
            self.coef_ = self.coef_ - (self.learning_rate * slope_of_coefficents )
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        return np.dot(X_test,self.coef_) + self.intercept_



## change the fit function code 

In [57]:
class StochasticDescantRegressor:

    def __init__(self,learning_rate=0.01,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept_ = 0
        self.coef_ = np.ones(X_train.shape[1])
        self.coef_ = np.ones(X_train.shape[1])
        
        for i in range(self.epoches):
            # update all the coeff and the intercept
            pass
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        return np.dot(X_test,self.coef_) + self.intercept_



tho huma har row ma update karna ha we run another loop

In [58]:
# example : generate a random integer numbe
np.random.randint(low = 0,high = X_train.shape[0])

303

In [59]:
# example random interger aya use them as row index
X_train[42]

array([ 0.0562386 , -0.04464164, -0.05794093, -0.00797713,  0.0520932 ,
        0.04910302,  0.05600338, -0.02141183, -0.02832317,  0.04448548])

In [60]:
coefficents = np.ones(X_train.shape[1]) # this is coefficnt matrix 
coefficents

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [61]:
# iska dot product karokeag self.coefficent ka sath

np.dot(X_train[42],coefficents)

0.09762897536168798

## slope of intercept

slope of the intercept = $\frac{\partial L}{\partial B_0} = \frac{-2}{n} \sum_{i=1}^{n}(y_i - \hat{y_i})$           
ya sari rows ka collectively nikalana ka liya tha           
ham sari rows ka upar kam nahi kar raha hum sirf ek row ka upar kam kar raha ha  tho now ab huma is summation ki koi jararut nahi ha $\sum_{i=1}^{n}$           
now ya m ka value ho jayega one now formula transform ka ho jayega
$\frac{\partial L}{\partial B_0} = -2 (y_i - \hat{y_i})$  ya par i ha apka row ka index     

In [62]:
class StochasticDescantRegressor:

    def __init__(self,learning_rate=0.01,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept_ = 0
        self.coef_ = np.ones(X_train.shape[1])
        self.coef_ = np.ones(X_train.shape[1])
        
        for i in range(self.epoches):
            for j in range(X_train.shapep[0]): # ya loop utni bar chalega jitni rows ha apka data ma 
                # generate a random number and use them as row index
                row_index = np.random.randint(low = 0,high = X_train.shape[0])
                
                # intercept update
                y_hat = np.dot(X_train[row_index],self.coef_) + self.intercept_
                
                slope_of_intercept  = -2 * (y_train[row_index] - y_hat)
                self.intercept_ = self.intercept_- (self.learning_rate * slope_of_intercept)
                # coefficents update 
               
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        return np.dot(X_test,self.coef_) + self.intercept_

## slope of coefficent 

slope of coeffient = $\frac{\partial L}{\partial B_1} = \frac{-2}{n} \sum_{i=1}^{n}(y_i - \hat{y_i})X_{i1}$         
ab hum is summation ki jarruat nahi ha ya na ka value 1 ho jayega kyuki ek row value ha ya jo Xi1 ha iski jaga huma ek row ka value chaya 
 $\frac{\partial L}{\partial B_1} = -2 (y_i - \hat{y_i})X_{i1}$             
 ya par hamra i ka matlab ha row index  

single row = (1,10)             
multiple by a scaler = (1,1)                
resulant ayega = 10 number              
(1,10) * (1,1) = 10 number              

In [63]:
class StochasticDescantRegressor:

    def __init__(self,learning_rate=0.01,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept_ = 0
        self.coef_ = np.ones(X_train.shape[1])
        self.coef_ = np.ones(X_train.shape[1])
        
        for i in range(self.epoches):
            for j in range(X_train.shape[0]): # ya loop utni bar chalega jitni rows ha apka data ma 
                # generate a random number and use them as row index
                row_index = np.random.randint(low = 0,high = X_train.shape[0])
                
                # intercept update
                y_hat = np.dot(X_train[row_index],self.coef_) + self.intercept_
                
                slope_of_intercept  = -2 * (y_train[row_index] - y_hat)
                self.intercept_ = self.intercept_- (self.learning_rate * slope_of_intercept)
                # coefficents update 
                slope_of_coefficents =  -2 * np.dot((y_train - y_hat),X_train)/X_train.shape[0]
                self.coef_ = self.coef_ - (self.learning_rate * slope_of_coefficents)
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        return np.dot(X_test,self.coef_) + self.intercept_
    
# creat object and fit the SGD regresion model
sgd_object = StochasticDescantRegressor(learning_rate=0.01,epoches=50)
sgd_object.fit(X_train,y_train)

intercept : 152.04422031577997
coefficents : [ 297.82958586   70.68679605  771.23504034  600.21107472  285.03527051
  227.82342493 -495.46726998  529.26473286  764.33338013  473.3431272 ]


In [64]:
# calcuate the predications
y_pred =sgd_object.predict(X_test)
#       actual - predication
r2_score(y_test,y_pred)

-1.3729936888860657

our score is so bad 

In [ ]:
    
# decrease the epoches  to 30
sgd_object = StochasticDescantRegressor(learning_rate=0.01,epoches=30)
sgd_object.fit(X_train,y_train)

intercept : 160.24270867781286
coefficents : [ 179.11155965   43.00696718  463.41882744  360.57351337  171.55154338
  137.36375276 -297.19335849  318.32447865  459.03990362  284.53618512]


In [67]:
# calcuate the predications
y_pred =sgd_object.predict(X_test)
#       actual - predication
r2_score(y_test,y_pred)

0.030896449010395677

wow 30 % varience is expained matlab thoda epoahoc increase karo 

In [ ]:
    
# increase the epoches to 45
sgd_object = StochasticDescantRegressor(learning_rate=0.01,epoches=45)
sgd_object.fit(X_train,y_train)

intercept : 152.88630341186618
coefficents : [ 268.14440368   63.68673238  694.16680022  540.28241977  256.61074753
  205.09765162 -445.77013955  476.37937886  687.99362075  426.08783019]


In [71]:
# calcuate the predications
y_pred =sgd_object.predict(X_test)
#       actual - predication
r2_score(y_test,y_pred)

-0.9060837535259372

Wow our model is best now it expalined the  90 % varience  of data 

In [73]:
    
# increase the epoches to 40
sgd_object = StochasticDescantRegressor(learning_rate=0.01,epoches=40)
sgd_object.fit(X_train,y_train)
# calcuate the predications and r2 score 
y_pred =sgd_object.predict(X_test)
print('r2score:' ,r2_score(y_test,y_pred))

intercept : 160.44478760757988
coefficents : [ 238.4960706    57.20676114  617.83992032  480.47884063  228.53416627
  183.09160792 -396.90828791  424.46978798  611.76027533  379.18028203]
r2score: -0.557805502406602


you can make a graph chanegineg epoachs value and r2 score 

# Time comparison
now check how much time for each gradient descant models 

In [74]:
import time

## Batch gradient Descant

In [93]:
class GradientDescantRegressor:

    def __init__(self,learning_rate=0.001,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept_ = 0
        self.coef_ = np.ones(X_train.shape[1])
        self.coef_ = np.ones(X_train.shape[1])
        
        for i in range(self.epoches):
            # update all the coeff and the intercept
            y_hat = np.dot(X_train,self.coef_) + self.intercept_
            slope_of_intercept  = -2 * np.mean(y_train - y_hat)
            self.intercept_ = self.intercept_- (self.learning_rate * slope_of_intercept) 

            slope_of_coefficents  =  -2 * np.dot((y_train - y_hat),X_train)/X_train.shape[0]
            self.coef_ = self.coef_ - (self.learning_rate * slope_of_coefficents )
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        pass


In [94]:
gd_object = GradientDescantRegressor()
start = time.time()
gd_object.fit(X_train,y_train)
print('time taken is :', time.time() - start)

intercept : 27.300032323395747
coefficents : [1.16382577 0.99950672 1.37912327 1.32841592 1.13278587 1.07264112
 0.78224268 1.22421099 1.42252845 1.23973457]
time taken is : 0.025972843170166016


In [ ]:

y_pred = gd_object.predict(X_test)
print('r2 score :', r2_score(y_test,y_pred))

## stochastic gradient descant

In [ ]:
class StochasticDescantRegressor:

    def __init__(self,learning_rate=0.01,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept_ = 0
        self.coef_ = np.ones(X_train.shape[1])
        self.coef_ = np.ones(X_train.shape[1])
        
        for i in range(self.epoches):
            for j in range(X_train.shape[0]): # ya loop utni bar chalega jitni rows ha apka data ma 
                # generate a random number and use them as row index
                row_index = np.random.randint(low = 0,high = X_train.shape[0])
                
                # intercept update
                y_hat = np.dot(X_train[row_index],self.coef_) + self.intercept_
                
                slope_of_intercept  = -2 * (y_train[row_index] - y_hat)
                self.intercept_ = self.intercept_- (self.learning_rate * slope_of_intercept)
                # coefficents update 
                slope_of_coefficents =  -2 * np.dot((y_train - y_hat),X_train)/X_train.shape[0]
                self.coef_ = self.coef_ - (self.learning_rate * slope_of_coefficents)
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        return np.dot(X_test,self.coef_) + self.intercept_
    


In [79]:

sgd_object = StochasticDescantRegressor(learning_rate=0.01,epoches=45)
start = time.time()
sgd_object.fit(X_train,y_train)
print('time taken it :', time.time() - start)

intercept : 163.12352904469338
coefficents : [ 268.15274657   63.80448493  694.33464933  540.31073781  256.68952395
  205.26060341 -445.95925248  476.60029842  688.01771361  426.16656295]
time taken it : 0.7962286472320557


batch  gradient descant time taken is : 0.025972843170166016           
stochastic gradient descant time taken it : 0.7962286472320557          
observation ; stochastic gradient descant ma jada time laga raha compare to batch  gradient descant

# SGD Regressor 

SGDRegressor iska andr internally stochastic gradient descant implement ha  iski achi baat ya ha ki ya alaga alag loss function ka upar kam kar sakta ha  isma jo SVM = support vector machine uska bhi loss function use kar sakta ho  it is a flexible class it work for other algorithums also.
- penailty as a regularisation apply kar sakta ho we study it in later
- alpha, l1 ratio is related to regularisation
- max_iter is epoches aap kitna iteration chalana chata ho
- tolerance aap kitna phala rokna chata ho ki 1000 epoches tak chata raha ki phala hi ruk jaya ki phala hi apko best solution mil jaya
    yadi pichala iteration ka value aor agala iteration ka value sa bhi kam ha tho aap aor aga kam nahi karoga aap aor aga update nahi karega
- shuffle = data ko shuffle karna ka liya
- epsilon  = related to other loss function
- random state = har bar apka stochastic gd alag answer dega yadi aap random state fix kar do atleast ek particular state ka liya apka answer same ayega
- learning rate = aap learning rate ko vary kar sakta ho
- eta = learning rate ka initaly value  


In [96]:
from sklearn.linear_model import SGDRegressor
print(SGDRegressor.__doc__)

Linear model fitted by minimizing a regularized empirical loss with SGD.

    SGD stands for Stochastic Gradient Descent: the gradient of the loss is
    estimated each sample at a time and the model is updated along the way with
    a decreasing strength schedule (aka learning rate).

    The regularizer is a penalty added to the loss function that shrinks model
    parameters towards the zero vector using either the squared euclidean norm
    L2 or the absolute norm L1 or a combination of both (Elastic Net). If the
    parameter update crosses the 0.0 value because of the regularizer, the
    update is truncated to 0.0 to allow for learning sparse models and achieve
    online feature selection.

    This implementation works with data represented as dense numpy arrays of
    floating point values for the features.

    Read more in the :ref:`User Guide <sgd>`.

    Parameters
    ----------
    loss : str, default='squared_error'
        The loss function to be used. The possible va

In [97]:
# now implent the class on our dataset
from sklearn.linear_model import SGDRegressor
sgd_model = SGDRegressor(max_iter=100,learning_rate='constant',eta0=0.01)
sgd_model.fit(X_train,y_train)

SGDRegressor(learning_rate='constant', max_iter=100)

In [99]:
print('intercept :',sgd_model.intercept_)
print('coefficents :',sgd_model.coef_)


intercept : [156.69059864]
coefficents : [  57.03106879  -61.76881094  339.60040432  241.04988766   20.90481897
  -22.87494421 -170.56266611  129.34150252  309.98904019  128.17336895]


In [101]:
y_pred = sgd_model.predict(X_test)
print('r2_score :',r2_score(y_test,y_pred))

r2_score : 0.42707499979545105


aap ek dataset lo uska upar SGD Regressor lagao parameter ka values change karo aor deko

Machine learning is all about performing experiments 
more experiment you do more learning you have 

jitna jada aap experiment perform karoga utni jada apko knowlege hoga